compiler/lib/global_deadcode.ml

```diff
@@ -0,0 +1,435 @@
+        (* If y is a live block, then x is the join of liveness fields that are x *)
+        | Live fields -> (
+            match defs.(Var.idx y) with
+            | Expr (Block (_, vars, _)) ->
+                let found = ref false in
+                Array.iteri
+                  ~f:(fun i v ->
+                    if Var.equal v x && IntSet.mem i fields then found := true)
+                  vars;
+                if !found then Top else Dead
+            | Expr (Field (_, i)) -> Live (IntSet.singleton i)
+            | _ -> Top)
+        (* If y is top and y is a field access, x depends only on that field *)
+        | Top -> (
+            match defs.(Var.idx y) with
+            | Expr (Field (_, i)) -> Live (IntSet.singleton i)
+            | _ -> Top))
+    (* If x is used as an argument for parameter y, then contribution is liveness of y *)
+    | Propagate -> Var.Tbl.get live_table y
+  in
+  Var.Map.fold
+    (fun y usage_kind live -> Domain.join (contribution y usage_kind) live)
+    uses.(idx)
+    (Domain.join live_vars.(idx) (Var.Tbl.get live_table x))
+
+let solver vars uses defs live_vars =
+  let g =
+    { G.domain = vars
+    ; G.iter_children = (fun f x -> Var.Map.iter (fun y _ -> f y) uses.(Var.idx x))
+    }
+  in
+  Solver.f () (G.invert () g) (propagate uses defs live_vars)
+
+(** Replace each instance of a dead variable with a sentinal value. 
+  Blocks that end in dead variables are compacted to the first live entry. 
+  Dead variables are replaced when
+    + They appear in a dead field of a block; or
+    + They are returned; or
+    + They are applied to a function. 
+ *)
+let zero prog sentinal live_table =
+  let compact_vars vars =
+    let i = ref (Array.length vars - 1) in
+    while !i >= 0 && Var.equal vars.(!i) sentinal do
+      i := !i - 1
+    done;
+    Array.sub vars ~pos:0 ~len:(!i + 1)
+  in
+  let is_live v =
+    match Var.Tbl.get live_table v with
+    | Dead -> false
+    | _ -> true
+  in
+  let zero_var x = if is_live x then x else sentinal in
+  let zero_instr instr =
+    match instr with
+    | Let (x, e) -> (
+        match e with
+        | Block (start, vars, is_array) -> (
+            match Var.Tbl.get live_table x with
+            | Live fields ->
+                let vars =
+                  Array.mapi
+                    ~f:(fun i v -> if IntSet.mem i fields then v else sentinal)
+                    vars
+                  |> compact_vars
+                in
+                let e = Block (start, vars, is_array) in
+                Let (x, e)
+            | _ -> instr)
+        | Apply ap ->
+            let args = List.map ~f:zero_var ap.args in
+            Let (x, Apply { ap with args })
+        | _ -> instr)
+    | _ -> instr
+  in
+  let zero_block block =
+    (* Analyze block instructions *)
+    let body = List.map ~f:(fun (instr, loc) -> zero_instr instr, loc) block.body in
+    (* Analyze branch *)
+    let branch =
+      let last, loc = block.branch in
```

Comment:
I would match on `block.branch` directly and return block.branch if it's not a `Return`

Reply:
[0214876](https://github.com/ocsigen/js_of_ocaml/pull/1503/commits/02148767e80b11152e311c0f437f8d5ba59f386f)